datalad/support/gitrepo.py

```diff
@@ -2355,17 +2358,67 @@ def get_submodules_(self, paths=None):
             return
 
         modinfo = self._parse_gitmodules()
-        for path, props in self.get_content_info(
-                paths=paths,
-                ref=None,
-                untracked='no').items():
-            if props.get('type', None) != 'dataset':
+        if not modinfo:
+            # we exit early, if there is nothing on record (even though
+            # a .gitmodules file exists).
+            # without this conditional exit, we would be able to discover
+            # subprojects even when they are not recorded in .gitmodules,
+            # but at the cost of running an unconstrained ls-files call
+            # below
+            return
+
+        # taken from 3.9's pathlib, can be removed once the minimally
+        # supported python version
+        def is_relative_to(self, *other):
+            """Return True if the path is relative to another path or False.
+            """
+            try:
+                self.relative_to(*other)
+                return True
+            except ValueError:
```

Comment:
I wonder if there is a guarantee that ValueError would be raised **only** if not relative, and not for some other (malformatted value etc) cases, thus possibly making it hard to debug.  I hope stdlib has lots of typing annotation and testing to guarantee that. Aforementioned code snippet might provide remedy.

Reply:
This is the implementation shipped with Python 3.9. What can be done to resolve your concern?

Reply:
concern is minor, but could be resolved through not using it, e.g. as proposed in https://github.com/datalad/datalad/pull/6942/files#r948226105 below